Installation of dependency 'gradunwarp' is error-prone

### Background

Currently, this project asks users to install [`gradunwarp`](https://github.com/Washington-University/gradunwarp/tree/v1.2.0) for gradient correction. However, `gradunwarp`'s installation methods are error-prone and unsafe:

* To help compile its own C extensions, `gradunwarp` uses [`numpy.distuils`](https://numpy.org/doc/stable/reference/distutils.html), rather than `setuptools` from the standard Python library. As a result, `pip install` can't be used. 
* Also, by default it [tries to install itself using the system Python](https://github.com/Washington-University/gradunwarp/tree/v1.2.0#install), which requires `sudo` privileges, or else this happens:

```
error: could not create '/usr/local/lib/python3.8/dist-packages/gradunwarp': Permission denied
```

### Issues with this

1. [Security risks](https://stackoverflow.com/a/21056000) related to running this installation as root.
2. In general, installing packages globally to the system Python is discouraged. ([[1]](https://packaging.python.org/guides/installing-using-pip-and-virtual-environments/#installing-virtualenv), [[2]](https://stackoverflow.com/a/22517157), [[3]](https://medium.com/better-programming/stop-installing-python-packages-globally-use-virtual-environments-a31dee9fb2de))
3. As an alternative to `sudo`, gradunwarp suggests using the `--prefix` switch to install `gradunwarp` somewhere else.
   * Users have to pick a location, then manually edit `PATH` and `PYTHONPATH`, which isn't very user-friendly. 
   * The flexibility here is error-prone. Troubleshooting issues requires us to understand where the user has chosen to install `gradunwarp` (which could be anywhere).
4. Regardless of where `gradunwarp` is installed, its dependencies (numpy, scipy, nibabel, etc.) are hard to keep track of. 
    * Ideally, the user could `pip install` gradunwarp and its dependencies inside a `virtualenv`, but this isn't possible.
    * Instead, gradunwarp [recommends](https://github.com/Washington-University/gradunwarp/tree/v1.2.0#prerequisites) using `sudo apt-get install python-numpy` and `sudo apt-get install python-scipy`. This installs the dependencies system-wide, which SciPy itself [lightly discourages](https://www.scipy.org/install.html#install-system-wide-via-a-package-manager).
    * In addition, the dependencies could be installed to a number of other locations and it would still work OK. Again, this flexibility is error-prone.

<details> 
<summary> Full list of locations where dependencies could potentially be installed </summary>

```console
joshua@XPS-15-9560$ python3 -m site
sys.path = [
    '/usr/lib/python38.zip',
    '/usr/lib/python3.8',
    '/usr/lib/python3.8/lib-dynload',
    '/home/joshua/.local/lib/python3.8/site-packages',
    '/usr/local/lib/python3.8/dist-packages',
    '/usr/lib/python3/dist-packages',
]
USER_BASE: '/home/joshua/.local' (exists)
USER_SITE: '/home/joshua/.local/lib/python3.8/site-packages' (exists)
ENABLE_USER_SITE: True
```
</details>

5. ~~If `gradunwarp`'s dependencies happen to be installed to the [user's `site-packages/` directory](https://stackoverflow.com/questions/31384639/what-is-pythons-site-packages-directory), then calling `gradient_unwarp.py` will fail inside `sct_run_batch`, because SCT has explicitly disabled user site-packages.~~ 
    **EDIT:** Fixed by https://github.com/neuropoly/spinalcordtoolbox/pull/3339.

### Recommendations

If possible, I think it would be good to come up with a better, more reliable way to install `gradunwarp`, then recommend that method in this project's README. 

(I almost want to say fork [(or copy)](https://github.com/Washington-University/gradunwarp/blob/v1.2.0/Copying.md) `gradunwarp` ourselves, then modify it so that `pip` and `virtualenv` be used.)

As I understand, `gradunwrap` isn't compatible with venvs.

I think the way to trick `gradunwrap` into behaving like a modern install is

```
python setup.py install --prefix=~/.local
```

On my system, this is on `PYTHONPATH` by default:

```
$ python3
Python 3.9.1 (default, Dec 13 2020, 11:55:53) 
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.path
['', '/usr/lib/python39.zip', '/usr/lib/python3.9', '/usr/lib/python3.9/lib-dynload', '/home/kousu/.local/lib/python3.9/site-packages', '/usr/lib/python3.9/site-packages']
>>>  # ^ /home/kousu/.local/lib/python3.9/site-packages
```

Now, it's *not* on python2's path:

```
$ python2
Python 2.7.18 (default, Sep  5 2020, 11:17:26) 
[GCC 10.2.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.path
['', '/usr/lib/python27.zip', '/usr/lib/python2.7', '/usr/lib/python2.7/plat-linux2', '/usr/lib/python2.7/lib-tk', '/usr/lib/python2.7/lib-old', '/usr/lib/python2.7/lib-dynload', '/usr/lib/python2.7/site-packages']
```

which maybe explains why they had to give the advice about manual installation back then.

<details><summary>experiment</summary>

I tried

```
$ python setup.py install --prefix=~/.local
```

<details><summary>compiling</summary>

```
running install
running build
running config_cc
unifing config_cc, config, build_clib, build_ext, build commands --compiler options
running config_fc
unifing config_fc, config, build_clib, build_ext, build commands --fcompiler options
running build_src
build_src
building py_modules sources
building extension "gradunwarp.core.interp3_ext" sources
building extension "gradunwarp.core.legendre_ext" sources
building extension "gradunwarp.core.transform_coordinates_ext" sources
building data_files sources
build_src: building npy-pkg config files
running build_py
package init file 'gradunwarp/core/tests/__init__.py' not found (or not a regular file)
creating build
creating build/lib.linux-x86_64-3.9
creating build/lib.linux-x86_64-3.9/gradunwarp
creating build/lib.linux-x86_64-3.9/gradunwarp/core
copying gradunwarp/core/__init__.py -> build/lib.linux-x86_64-3.9/gradunwarp/core
copying gradunwarp/core/coeffs.py -> build/lib.linux-x86_64-3.9/gradunwarp/core
copying gradunwarp/core/globals.py -> build/lib.linux-x86_64-3.9/gradunwarp/core
copying gradunwarp/__init__.py -> build/lib.linux-x86_64-3.9/gradunwarp
copying gradunwarp/core/utils.py -> build/lib.linux-x86_64-3.9/gradunwarp/core
copying gradunwarp/core/unwarp_resample.py -> build/lib.linux-x86_64-3.9/gradunwarp/core
copying gradunwarp/core/gradient_unwarp.py -> build/lib.linux-x86_64-3.9/gradunwarp/core
creating build/lib.linux-x86_64-3.9/gradunwarp/core/tests
copying gradunwarp/core/tests/test_utils.py -> build/lib.linux-x86_64-3.9/gradunwarp/core/tests
package init file 'gradunwarp/core/tests/__init__.py' not found (or not a regular file)
warning: build_py: byte-compiling is disabled, skipping.

running build_ext
customize UnixCCompiler
customize UnixCCompiler using build_ext
CCompilerOpt.cc_test_flags[999] : testing flags (-march=native)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

creating /tmp/tmppdoxc9zn/usr
creating /tmp/tmppdoxc9zn/usr/lib
creating /tmp/tmppdoxc9zn/usr/lib/python3.9
creating /tmp/tmppdoxc9zn/usr/lib/python3.9/site-packages
creating /tmp/tmppdoxc9zn/usr/lib/python3.9/site-packages/numpy
creating /tmp/tmppdoxc9zn/usr/lib/python3.9/site-packages/numpy/distutils
creating /tmp/tmppdoxc9zn/usr/lib/python3.9/site-packages/numpy/distutils/checks
compile options: '-I/usr/include/python3.9 -c'
extra options: '-march=native'
CCompilerOpt.cc_test_flags[999] : testing flags (-O3)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-O3'
CCompilerOpt.cc_test_flags[999] : testing flags (-Werror)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-Werror'
CCompilerOpt.__init__[1674] : check requested baseline
CCompilerOpt.cc_test_flags[999] : testing flags (-msse)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse'
CCompilerOpt.cc_test_flags[999] : testing flags (-msse2)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse2'
CCompilerOpt.feature_test[1444] : testing feature 'SSE' with flags (-msse -msse2)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -Werror'
CCompilerOpt.feature_test[1444] : testing feature 'SSE2' with flags (-msse -msse2)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -Werror'
CCompilerOpt.cc_test_flags[999] : testing flags (-msse3)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse3'
CCompilerOpt.feature_test[1444] : testing feature 'SSE3' with flags (-msse -msse2 -msse3)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -Werror'
CCompilerOpt.__init__[1683] : check requested dispatch-able features
CCompilerOpt.cc_test_flags[999] : testing flags (-mssse3)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-mssse3'
CCompilerOpt.cc_test_flags[999] : testing flags (-msse4.1)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse4.1'
CCompilerOpt.feature_test[1444] : testing feature 'SSE41' with flags (-msse -msse2 -msse3 -mssse3 -msse4.1)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -Werror'
CCompilerOpt.feature_test[1444] : testing feature 'SSSE3' with flags (-msse -msse2 -msse3 -mssse3)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -Werror'
CCompilerOpt.cc_test_flags[999] : testing flags (-mpopcnt)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-mpopcnt'
CCompilerOpt.cc_test_flags[999] : testing flags (-msse4.2)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse4.2'
CCompilerOpt.feature_test[1444] : testing feature 'SSE42' with flags (-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -Werror'
CCompilerOpt.feature_test[1444] : testing feature 'POPCNT' with flags (-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -Werror'
CCompilerOpt.cc_test_flags[999] : testing flags (-mavx)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-mavx'
CCompilerOpt.feature_test[1444] : testing feature 'AVX' with flags (-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -Werror'
CCompilerOpt.cc_test_flags[999] : testing flags (-mf16c)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-mf16c'
CCompilerOpt.cc_test_flags[999] : testing flags (-mfma)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-mfma'
CCompilerOpt.cc_test_flags[999] : testing flags (-mavx2)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-mavx2'
CCompilerOpt.cc_test_flags[999] : testing flags (-mavx512f)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-mavx512f'
CCompilerOpt.feature_test[1444] : testing feature 'AVX512F' with flags (-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -Werror'
CCompilerOpt.feature_test[1444] : testing feature 'F16C' with flags (-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -Werror'
CCompilerOpt.cc_test_flags[999] : testing flags (-mavx512cd)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-mavx512cd'
CCompilerOpt.cc_test_flags[999] : testing flags (-mavx512er -mavx512pf)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-mavx512er -mavx512pf'
CCompilerOpt.feature_test[1444] : testing feature 'AVX512_KNL' with flags (-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512er -mavx512pf)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512er -mavx512pf -Werror'
CCompilerOpt.feature_test[1444] : testing feature 'AVX2' with flags (-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mavx2)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mavx2 -Werror'
CCompilerOpt.feature_test[1444] : testing feature 'FMA3' with flags (-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -Werror'
CCompilerOpt.feature_test[1444] : testing feature 'AVX512CD' with flags (-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -Werror'
CCompilerOpt.cc_test_flags[999] : testing flags (-mavx5124fmaps -mavx5124vnniw -mavx512vpopcntdq)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-mavx5124fmaps -mavx5124vnniw -mavx512vpopcntdq'
CCompilerOpt.feature_test[1444] : testing feature 'AVX512_KNM' with flags (-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512er -mavx512pf -mavx5124fmaps -mavx5124vnniw -mavx512vpopcntdq)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512er -mavx512pf -mavx5124fmaps -mavx5124vnniw -mavx512vpopcntdq -Werror'
CCompilerOpt.cc_test_flags[999] : testing flags (-mavx512vl -mavx512bw -mavx512dq)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-mavx512vl -mavx512bw -mavx512dq'
CCompilerOpt.feature_test[1444] : testing feature 'AVX512_SKX' with flags (-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512vl -mavx512bw -mavx512dq)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512vl -mavx512bw -mavx512dq -Werror'
CCompilerOpt.cc_test_flags[999] : testing flags (-mavx512vnni)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-mavx512vnni'
CCompilerOpt.feature_test[1444] : testing feature 'AVX512_CLX' with flags (-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512vl -mavx512bw -mavx512dq -mavx512vnni)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512vl -mavx512bw -mavx512dq -mavx512vnni -Werror'
CCompilerOpt.cc_test_flags[999] : testing flags (-mavx512ifma -mavx512vbmi)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-mavx512ifma -mavx512vbmi'
CCompilerOpt.feature_test[1444] : testing feature 'AVX512_CNL' with flags (-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512vl -mavx512bw -mavx512dq -mavx512ifma -mavx512vbmi)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512vl -mavx512bw -mavx512dq -mavx512ifma -mavx512vbmi -Werror'
CCompilerOpt.cc_test_flags[999] : testing flags (-mavx512vbmi2 -mavx512bitalg -mavx512vpopcntdq)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-mavx512vbmi2 -mavx512bitalg -mavx512vpopcntdq'
CCompilerOpt.feature_test[1444] : testing feature 'AVX512_ICL' with flags (-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512vl -mavx512bw -mavx512dq -mavx512vnni -mavx512ifma -mavx512vbmi -mavx512vbmi2 -mavx512bitalg -mavx512vpopcntdq)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512vl -mavx512bw -mavx512dq -mavx512vnni -mavx512ifma -mavx512vbmi -mavx512vbmi2 -mavx512bitalg -mavx512vpopcntdq -Werror'
CCompilerOpt.__init__[1695] : skip features (SSE3 SSE SSE2) since its part of baseline
CCompilerOpt.__init__[1699] : initialize targets groups
CCompilerOpt.__init__[1701] : parse target group simd_test
CCompilerOpt._parse_target_tokens[1910] : skip targets (FMA4 VSX2 VSX3 NEON VSX ASIMD XOP) not part of baseline or dispatch-able features
CCompilerOpt._parse_policy_not_keepbase[2022] : skip baseline features (SSE2)
CCompilerOpt.generate_dispatch_header[2236] : generate CPU dispatch header: (build/src.linux-x86_64-3.9/numpy/distutils/include/npy_cpu_dispatch_config.h)
CCompilerOpt.generate_dispatch_header[2245] : dispatch header dir build/src.linux-x86_64-3.9/numpy/distutils/include does not exist, creating it
CCompilerOpt.feature_extra_checks[1519] : Testing extra checks for feature 'AVX512F' (AVX512F_REDUCE)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -Werror'
CCompilerOpt.feature_extra_checks[1519] : Testing extra checks for feature 'AVX512_SKX' (AVX512BW_MASK AVX512DQ_MASK)
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512vl -mavx512bw -mavx512dq -Werror'
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/include/python3.9 -c'
extra options: '-msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512vl -mavx512bw -mavx512dq -Werror'
Detected changes on compiler optimizations, force rebuilding
building 'gradunwarp.core.interp3_ext' extension
compiling C sources
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

creating build/temp.linux-x86_64-3.9/gradunwarp
creating build/temp.linux-x86_64-3.9/gradunwarp/core
compile options: '-I/usr/lib/python3.9/site-packages/numpy/core/include -I/usr/lib/python3.9/site-packages/numpy/core/include -Ibuild/src.linux-x86_64-3.9/numpy/distutils/include -I/usr/include/python3.9 -c'
extra options: '-O3 -msse -msse2 -msse3'
gcc: gradunwarp/core/interp3_ext.c
 In file included from /usr/lib/python3.9/site-packages/numpy/core/include/numpy/ndarraytypes.h:1944,
                 from /usr/lib/python3.9/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                 from /usr/lib/python3.9/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from gradunwarp/core/interp3_ext.c:10:
/usr/lib/python3.9/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
   17 | #warning "Using deprecated NumPy API, disable it with " \
      |  ^~~~~~~
gradunwarp/core/interp3_ext.c: In function ‘interp3’:
gradunwarp/core/interp3_ext.c:97:51: warning: passing argument 1 of ‘(PyObject * (*)(PyObject *))*(PyArray_API + 784)’ from incompatible pointer type [-Wincompatible-pointer-types]
   97 |     itr_v = (PyArrayIterObject *) PyArray_IterNew(result);
      |                                                   ^~~~~~
      |                                                   |
      |                                                   PyArrayObject * {aka struct tagPyArrayObject_fields *}
gradunwarp/core/interp3_ext.c:97:51: note: expected ‘PyObject *’ {aka ‘struct _object *’} but argument is of type ‘PyArrayObject *’ {aka ‘struct tagPyArrayObject_fields *’}
gradunwarp/core/interp3_ext.c:98:51: warning: passing argument 1 of ‘(PyObject * (*)(PyObject *))*(PyArray_API + 784)’ from incompatible pointer type [-Wincompatible-pointer-types]
   98 |     itr_r = (PyArrayIterObject *) PyArray_IterNew(R);
      |                                                   ^
      |                                                   |
      |                                                   PyArrayObject * {aka struct tagPyArrayObject_fields *}
gradunwarp/core/interp3_ext.c:98:51: note: expected ‘PyObject *’ {aka ‘struct _object *’} but argument is of type ‘PyArrayObject *’ {aka ‘struct tagPyArrayObject_fields *’}
gradunwarp/core/interp3_ext.c:99:51: warning: passing argument 1 of ‘(PyObject * (*)(PyObject *))*(PyArray_API + 784)’ from incompatible pointer type [-Wincompatible-pointer-types]
   99 |     itr_c = (PyArrayIterObject *) PyArray_IterNew(C);
      |                                                   ^
      |                                                   |
      |                                                   PyArrayObject * {aka struct tagPyArrayObject_fields *}
gradunwarp/core/interp3_ext.c:99:51: note: expected ‘PyObject *’ {aka ‘struct _object *’} but argument is of type ‘PyArrayObject *’ {aka ‘struct tagPyArrayObject_fields *’}
gradunwarp/core/interp3_ext.c:100:51: warning: passing argument 1 of ‘(PyObject * (*)(PyObject *))*(PyArray_API + 784)’ from incompatible pointer type [-Wincompatible-pointer-types]
  100 |     itr_s = (PyArrayIterObject *) PyArray_IterNew(S);
      |                                                   ^
      |                                                   |
      |                                                   PyArrayObject * {aka struct tagPyArrayObject_fields *}
gradunwarp/core/interp3_ext.c:100:51: note: expected ‘PyObject *’ {aka ‘struct _object *’} but argument is of type ‘PyArrayObject *’ {aka ‘struct tagPyArrayObject_fields *’}
gradunwarp/core/interp3_ext.c:123:9: warning: returning ‘PyArrayObject *’ {aka ‘struct tagPyArrayObject_fields *’} from a function with incompatible return type ‘PyObject *’ {aka ‘struct _object *’} [-Wincompatible-pointer-types]
  123 |  return result;
      |         ^~~~~~
gcc -pthread -shared -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -fno-semantic-interposition -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now build/temp.linux-x86_64-3.9/gradunwarp/core/interp3_ext.o -L/usr/lib -o build/lib.linux-x86_64-3.9/gradunwarp/core/interp3_ext.cpython-39-x86_64-linux-gnu.so
building 'gradunwarp.core.legendre_ext' extension
compiling C sources
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/lib/python3.9/site-packages/numpy/core/include -I/usr/lib/python3.9/site-packages/numpy/core/include -Ibuild/src.linux-x86_64-3.9/numpy/distutils/include -I/usr/include/python3.9 -c'
extra options: '-O3 -msse -msse2 -msse3'
gcc: gradunwarp/core/legendre_ext.c
In file included from /usr/lib/python3.9/site-packages/numpy/core/include/numpy/ndarraytypes.h:1944,
                 from /usr/lib/python3.9/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                 from /usr/lib/python3.9/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from gradunwarp/core/legendre_ext.c:10:
/usr/lib/python3.9/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
   17 | #warning "Using deprecated NumPy API, disable it with " \
      |  ^~~~~~~
gradunwarp/core/legendre_ext.c: In function ‘legendre’:
gradunwarp/core/legendre_ext.c:86:25: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘long int’ [-Wformat=]
   86 |         printf("but mu=%d and nu=%d\n", mu, nu);
      |                        ~^               ~~
      |                         |               |
      |                         int             long int
      |                        %ld
gradunwarp/core/legendre_ext.c:86:35: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘long int’ [-Wformat=]
   86 |         printf("but mu=%d and nu=%d\n", mu, nu);
      |                                  ~^         ~~
      |                                   |         |
      |                                   int       long int
      |                                  %ld
gradunwarp/core/legendre_ext.c:99:51: warning: passing argument 1 of ‘(PyObject * (*)(PyObject *))*(PyArray_API + 784)’ from incompatible pointer type [-Wincompatible-pointer-types]
   99 |     itr_x = (PyArrayIterObject *) PyArray_IterNew(x);
      |                                                   ^
      |                                                   |
      |                                                   PyArrayObject * {aka struct tagPyArrayObject_fields *}
gradunwarp/core/legendre_ext.c:99:51: note: expected ‘PyObject *’ {aka ‘struct _object *’} but argument is of type ‘PyArrayObject *’ {aka ‘struct tagPyArrayObject_fields *’}
gradunwarp/core/legendre_ext.c:100:51: warning: passing argument 1 of ‘(PyObject * (*)(PyObject *))*(PyArray_API + 784)’ from incompatible pointer type [-Wincompatible-pointer-types]
  100 |     itr_r = (PyArrayIterObject *) PyArray_IterNew(result);
      |                                                   ^~~~~~
      |                                                   |
      |                                                   PyArrayObject * {aka struct tagPyArrayObject_fields *}
gradunwarp/core/legendre_ext.c:100:51: note: expected ‘PyObject *’ {aka ‘struct _object *’} but argument is of type ‘PyArrayObject *’ {aka ‘struct tagPyArrayObject_fields *’}
gradunwarp/core/legendre_ext.c:108:32: warning: format ‘%f’ expects argument of type ‘double’, but argument 2 has type ‘PyArrayObject *’ {aka ‘struct tagPyArrayObject_fields *’} [-Wformat=]
  108 |             printf("but got x=%f\n", x);
      |                               ~^     ~
      |                                |     |
      |                                |     PyArrayObject * {aka struct tagPyArrayObject_fields *}
      |                                double
gradunwarp/core/legendre_ext.c:117:12: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]
  117 |            if ( mu & 1)
      |            ^~
gradunwarp/core/legendre_ext.c:119:15: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘if’
  119 |               p_nu = s * odd_factorial(2 * mu - 1) * pow(sqrt( 1.0 - *px * *px), mu);
      |               ^~~~
gradunwarp/core/legendre_ext.c:157:12: warning: returning ‘PyArrayObject *’ {aka ‘struct tagPyArrayObject_fields *’} from a function with incompatible return type ‘PyObject *’ {aka ‘struct _object *’} [-Wincompatible-pointer-types]
  157 |     return result;
      |            ^~~~~~
gcc -pthread -shared -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -fno-semantic-interposition -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now build/temp.linux-x86_64-3.9/gradunwarp/core/legendre_ext.o -L/usr/lib -o build/lib.linux-x86_64-3.9/gradunwarp/core/legendre_ext.cpython-39-x86_64-linux-gnu.so
building 'gradunwarp.core.transform_coordinates_ext' extension
compiling C sources
C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fno-semantic-interposition -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC

compile options: '-I/usr/lib/python3.9/site-packages/numpy/core/include -I/usr/lib/python3.9/site-packages/numpy/core/include -Ibuild/src.linux-x86_64-3.9/numpy/distutils/include -I/usr/include/python3.9 -c'
extra options: '-O3 -msse -msse2 -msse3'
gcc: gradunwarp/core/transform_coordinates_ext.c
In file included from /usr/lib/python3.9/site-packages/numpy/core/include/numpy/ndarraytypes.h:1944,
                 from /usr/lib/python3.9/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                 from /usr/lib/python3.9/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from gradunwarp/core/transform_coordinates_ext.c:10:
/usr/lib/python3.9/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
   17 | #warning "Using deprecated NumPy API, disable it with " \
      |  ^~~~~~~
gradunwarp/core/transform_coordinates_ext.c: In function ‘transform_coordinates’:
gradunwarp/core/transform_coordinates_ext.c:102:51: warning: passing argument 1 of ‘(PyObject * (*)(PyObject *))*(PyArray_API + 784)’ from incompatible pointer type [-Wincompatible-pointer-types]
  102 |     itr_x = (PyArrayIterObject *) PyArray_IterNew(X);
      |                                                   ^
      |                                                   |
      |                                                   PyArrayObject * {aka struct tagPyArrayObject_fields *}
gradunwarp/core/transform_coordinates_ext.c:102:51: note: expected ‘PyObject *’ {aka ‘struct _object *’} but argument is of type ‘PyArrayObject *’ {aka ‘struct tagPyArrayObject_fields *’}
gradunwarp/core/transform_coordinates_ext.c:103:51: warning: passing argument 1 of ‘(PyObject * (*)(PyObject *))*(PyArray_API + 784)’ from incompatible pointer type [-Wincompatible-pointer-types]
  103 |     itr_y = (PyArrayIterObject *) PyArray_IterNew(Y);
      |                                                   ^
      |                                                   |
      |                                                   PyArrayObject * {aka struct tagPyArrayObject_fields *}
gradunwarp/core/transform_coordinates_ext.c:103:51: note: expected ‘PyObject *’ {aka ‘struct _object *’} but argument is of type ‘PyArrayObject *’ {aka ‘struct tagPyArrayObject_fields *’}
gradunwarp/core/transform_coordinates_ext.c:104:51: warning: passing argument 1 of ‘(PyObject * (*)(PyObject *))*(PyArray_API + 784)’ from incompatible pointer type [-Wincompatible-pointer-types]
  104 |     itr_z = (PyArrayIterObject *) PyArray_IterNew(Z);
      |                                                   ^
      |                                                   |
      |                                                   PyArrayObject * {aka struct tagPyArrayObject_fields *}
gradunwarp/core/transform_coordinates_ext.c:104:51: note: expected ‘PyObject *’ {aka ‘struct _object *’} but argument is of type ‘PyArrayObject *’ {aka ‘struct tagPyArrayObject_fields *’}
gradunwarp/core/transform_coordinates_ext.c:105:52: warning: passing argument 1 of ‘(PyObject * (*)(PyObject *))*(PyArray_API + 784)’ from incompatible pointer type [-Wincompatible-pointer-types]
  105 |     itr_xm = (PyArrayIterObject *) PyArray_IterNew(Xm);
      |                                                    ^~
      |                                                    |
      |                                                    PyArrayObject * {aka struct tagPyArrayObject_fields *}
gradunwarp/core/transform_coordinates_ext.c:105:52: note: expected ‘PyObject *’ {aka ‘struct _object *’} but argument is of type ‘PyArrayObject *’ {aka ‘struct tagPyArrayObject_fields *’}
gradunwarp/core/transform_coordinates_ext.c:106:52: warning: passing argument 1 of ‘(PyObject * (*)(PyObject *))*(PyArray_API + 784)’ from incompatible pointer type [-Wincompatible-pointer-types]
  106 |     itr_ym = (PyArrayIterObject *) PyArray_IterNew(Ym);
      |                                                    ^~
      |                                                    |
      |                                                    PyArrayObject * {aka struct tagPyArrayObject_fields *}
gradunwarp/core/transform_coordinates_ext.c:106:52: note: expected ‘PyObject *’ {aka ‘struct _object *’} but argument is of type ‘PyArrayObject *’ {aka ‘struct tagPyArrayObject_fields *’}
gradunwarp/core/transform_coordinates_ext.c:107:52: warning: passing argument 1 of ‘(PyObject * (*)(PyObject *))*(PyArray_API + 784)’ from incompatible pointer type [-Wincompatible-pointer-types]
  107 |     itr_zm = (PyArrayIterObject *) PyArray_IterNew(Zm);
      |                                                    ^~
      |                                                    |
      |                                                    PyArrayObject * {aka struct tagPyArrayObject_fields *}
gradunwarp/core/transform_coordinates_ext.c:107:52: note: expected ‘PyObject *’ {aka ‘struct _object *’} but argument is of type ‘PyArrayObject *’ {aka ‘struct tagPyArrayObject_fields *’}
gradunwarp/core/transform_coordinates_ext.c:73:12: warning: unused variable ‘m’ [-Wunused-variable]
   73 |     float *m;
      |            ^
gcc -pthread -shared -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -fno-semantic-interposition -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now build/temp.linux-x86_64-3.9/gradunwarp/core/transform_coordinates_ext.o -L/usr/lib -o build/lib.linux-x86_64-3.9/gradunwarp/core/transform_coordinates_ext.cpython-39-x86_64-linux-gnu.so
running build_scripts
creating build/scripts.linux-x86_64-3.9
copying and adjusting gradunwarp/core/gradient_unwarp.py -> build/scripts.linux-x86_64-3.9
changing mode of build/scripts.linux-x86_64-3.9/gradient_unwarp.py from 644 to 755
running install_lib
creating /home/kousu/.local/lib/python3.9/site-packages/gradunwarp
copying build/lib.linux-x86_64-3.9/gradunwarp/__init__.py -> /home/kousu/.local/lib/python3.9/site-packages/gradunwarp
creating /home/kousu/.local/lib/python3.9/site-packages/gradunwarp/core
copying build/lib.linux-x86_64-3.9/gradunwarp/core/__init__.py -> /home/kousu/.local/lib/python3.9/site-packages/gradunwarp/core
copying build/lib.linux-x86_64-3.9/gradunwarp/core/legendre_ext.cpython-39-x86_64-linux-gnu.so -> /home/kousu/.local/lib/python3.9/site-packages/gradunwarp/core
copying build/lib.linux-x86_64-3.9/gradunwarp/core/globals.py -> /home/kousu/.local/lib/python3.9/site-packages/gradunwarp/core
copying build/lib.linux-x86_64-3.9/gradunwarp/core/interp3_ext.cpython-39-x86_64-linux-gnu.so -> /home/kousu/.local/lib/python3.9/site-packages/gradunwarp/core
creating /home/kousu/.local/lib/python3.9/site-packages/gradunwarp/core/tests
copying build/lib.linux-x86_64-3.9/gradunwarp/core/tests/test_utils.py -> /home/kousu/.local/lib/python3.9/site-packages/gradunwarp/core/tests
copying build/lib.linux-x86_64-3.9/gradunwarp/core/unwarp_resample.py -> /home/kousu/.local/lib/python3.9/site-packages/gradunwarp/core
copying build/lib.linux-x86_64-3.9/gradunwarp/core/utils.py -> /home/kousu/.local/lib/python3.9/site-packages/gradunwarp/core
copying build/lib.linux-x86_64-3.9/gradunwarp/core/transform_coordinates_ext.cpython-39-x86_64-linux-gnu.so -> /home/kousu/.local/lib/python3.9/site-packages/gradunwarp/core
copying build/lib.linux-x86_64-3.9/gradunwarp/core/gradient_unwarp.py -> /home/kousu/.local/lib/python3.9/site-packages/gradunwarp/core
copying build/lib.linux-x86_64-3.9/gradunwarp/core/coeffs.py -> /home/kousu/.local/lib/python3.9/site-packages/gradunwarp/core
warning: install_lib: byte-compiling is disabled, skipping.

running install_scripts
copying build/scripts.linux-x86_64-3.9/gradient_unwarp.py -> /home/kousu/.local/bin
changing mode of /home/kousu/.local/bin/gradient_unwarp.py to 755
running install_data
copying gradunwarp/core/transform_coordinates_ext.c -> /home/kousu/.local/lib/python3.9/site-packages/gradunwarp/core/
copying gradunwarp/core/legendre_ext.c -> /home/kousu/.local/lib/python3.9/site-packages/gradunwarp/core/
copying gradunwarp/core/interp3_ext.c -> /home/kousu/.local/lib/python3.9/site-packages/gradunwarp/core/
running install_egg_info
Writing /home/kousu/.local/lib/python3.9/site-packages/gradunwarp-HCP_1.2.0-py3.9.egg-info
running install_clib
customize UnixCCompiler

########### EXT COMPILER OPTIMIZATION ###########
Platform      : 
  Architecture: x64
  Compiler    : gcc

CPU baseline  : 
  Requested   : 'min'
  Enabled     : SSE SSE2 SSE3
  Flags       : -msse -msse2 -msse3
  Extra checks: none

CPU dispatch  : 
  Requested   : 'max -xop -fma4'
  Enabled     : SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2 AVX512F AVX512CD AVX512_KNL AVX512_KNM AVX512_SKX AVX512_CLX AVX512_CNL AVX512_ICL
  Generated   : none
CCompilerOpt._cache_write[796] : write cache to path -> /home/kousu/src/neuropoly/gradunwarp/build/temp.linux-x86_64-3.9/ccompiler_opt_cache_ext.py
```

</details>

```
$ gradient_unwarp.py 
Traceback (most recent call last):
  File "/home/kousu/.local/bin/gradient_unwarp.py", line 11, in <module>
    from gradunwarp.core import (globals, coeffs, utils)
  File "/home/kousu/.local/lib/python3.9/site-packages/gradunwarp/core/__init__.py", line 8, in <module>
    from .unwarp_resample import Unwarper
  File "/home/kousu/.local/lib/python3.9/site-packages/gradunwarp/core/unwarp_resample.py", line 20, in <module>
    import nibabel as nib
ModuleNotFoundError: No module named 'nibabel'
$ command -v gradient_unwarp.py
/home/kousu/.local/bin/gradient_unwarp.py
```

So it almost works, it's just that it doesn't install its own dependencies; and I tried

```diff
diff --git a/setup.py b/setup.py
index 3631111..6f76b5e 100644
--- a/setup.py
+++ b/setup.py
@@ -46,6 +46,7 @@ setup(name='gradunwarp',
       author = 'Human Connectome Project',
       py_modules  = mods,
       ext_modules = [ext1, ext2, ext3],
+      install_requires=['nibabel'],
       scripts = scripts_cmd,
       configuration=configuration,
      )
```

but this just got

```
$ python setup.py install --prefix=~/.local
/usr/lib/python3.9/distutils/dist.py:274: UserWarning: Unknown distribution option: 'install_requires'
  warnings.warn(msg)
```

because it's numpy.distutils, not the one that understand wheels.

</details>

I think setting `PYTHONUSERSITE=True` and other things like it is not a good solution. Too many apps, with total reasonability IMO, assume that they are running on a standard-ish system, and breaking those assumptions makes SCT an incompatible thing like Matlab that people have to containerize to keep running. If we're going to force people to containerize we should at least provide the docker image ourselves.

We could try to isolate `PYTHONUSERSITE` to just SCT itself, maybe? Like, make sure every: `subprocess.run([...], env={k: os.environ[k] for k in set(os.environ.keys()) - {'PYTHONUSERSITE'}})`. That would be a hacky workaround but at least it wouldn't leak out beyond SCT anymore.

I'm sure we could fix gradunwarp to be a modern pip package with wheels available. We just have to go look at what sklearn or scikit-image have in their build scripts.

--- a/README.md
+++ b/README.md
@@ -87,7 +87,7 @@ This repository includes a collection of scripts to analyse a BIDS-structured MR
 MANDATORY:
 
 * [SCT 5.3.0](https://github.com/neuropoly/spinalcordtoolbox/releases/tag/5.3.0) for processing
-* [gradunwrap v1.2.0](https://github.com/Washington-University/gradunwarp/tree/v1.2.0) for gradient correction
+* [gradunwarp](gradunwarp_installation.md) for gradient correction
 * [ANTs](https://github.com/ANTsX/ANTs/wiki/Compiling-ANTs-on-Linux-and-Mac-OS) for processing
 * Python 3.7  for statistical analysis
 
@@ -111,7 +111,7 @@ cd ukbiobank-spinalcord-csa
 pip install -e ./
 ~~~
 #### Note on gradient distortion correction
-A `coeff.grad` associated with the MRI scanner used for the data is necessary if it has not been applied yet. In this project, the gradient distortion correction is done in `preprocess_data.sh` with [gradunwrap v1.2.0](https://github.com/Washington-University/gradunwarp/tree/v1.2.0) and Siemens `coeff.grad` file.
+A `coeff.grad` associated with the MRI scanner used for the data is necessary if it has not been applied yet. In this project, the gradient distortion correction is done in `preprocess_data.sh` with `gradunwarp` and Siemens `coeff.grad` file.
 - - -
 ### Preprocessing
 Preprocessing generates a dataset with gradient distortion correction. 

--- a/README.md
+++ b/README.md
@@ -86,7 +86,7 @@ This repository includes a collection of scripts to analyse a BIDS-structured MR
 ### Dependencies
 MANDATORY:
 
-* [SCT 5.0.1](https://github.com/neuropoly/spinalcordtoolbox/releases/tag/5.0.1) for processing
+* [SCT 5.3.0](https://github.com/neuropoly/spinalcordtoolbox/releases/tag/5.3.0) for processing
 * [gradunwrap v1.2.0](https://github.com/Washington-University/gradunwarp/tree/v1.2.0) for gradient correction
 * [ANTs](https://github.com/ANTsX/ANTs/wiki/Compiling-ANTs-on-Linux-and-Mac-OS) for processing
 * Python 3.7  for statistical analysis

--- a/gradunwarp_installation.md
+++ b/gradunwarp_installation.md
@@ -1,0 +1,33 @@
+### How to install gradunwarp
+
+We recommend that you **do not** follow the default installation instructions for `gradunwarp`, due to the reasons outlined [here](https://github.com/sct-pipeline/ukbiobank-spinalcord-csa/issues/31).
+
+Instead, we provide our own set of steps to help make the installation process smoother.
+
+1. Install the dependencies of `gradunwarp`
+
+```bash
+pip install numpy scipy nibabel
+```
+
+2. Download and extract the [latest release](https://github.com/Washington-University/gradunwarp/releases/latest) of `gradunwarp`
+   
+
+3. Navigate to the extracted folder, then install `gradunwarp`
+
+```bash
+cd gradunwarp-1.2.0/
+python setup.py install --prefix="~/.local"
+```
+
+4. Update your `$PATH` to include the installation location
+
+```bash
+# For bash users
+echo "export PATH=$PATH:~/.local/bin" >> ~/.bashrc
+
+# For zsh users
+echo "export PATH=$PATH:~/.local/bin" >> ~/.zshrc
+```
+   
+5. You're done! The command `gradient_unwarp.py` should now be available in your terminal after restarting the terminal.

